src/main/scala/com/banzaicloud/spark/metrics/sink/PrometheusSink.scala

```diff
@@ -17,28 +17,32 @@
 package com.banzaicloud.spark.metrics.sink
 
 import java.io.File
-import java.net.InetAddress
-import java.net.URI
-import java.net.UnknownHostException
+import java.net.{InetAddress, URI, UnknownHostException}
 import java.util
 import java.util.Properties
 import java.util.concurrent.TimeUnit
 
-import com.banzaicloud.metrics.prometheus.client.exporter.PushGatewayWithTimestamp
-import com.banzaicloud.spark.metrics.DropwizardExportsWithMetricNameCaptureAndReplace
+import com.banzaicloud.spark.metrics.NameDecorator.Replace
+import com.banzaicloud.spark.metrics.{SparkDropwizardExports, SparkJmxExports}
 import com.codahale.metrics._
 import io.prometheus.client.CollectorRegistry
 import io.prometheus.client.dropwizard.DropwizardExports
-import org.apache.spark.internal.Logging
+import io.prometheus.client.exporter.PushGateway
```

Comment:
This doesn't support passing timestamps. There are older and custom versions of PushGateway that support timestamps. We need to support those that's why we implemented `PushGatewayWithTimestamp`

Reply:
@stoader Why do you think so? TextFormat class has `if` for timestamp: 
https://github.com/prometheus/client_java/blob/master/simpleclient_common/src/main/java/io/prometheus/client/exporter/common/TextFormat.java#L49

I'll try to write test for that

Reply:
https://github.com/prometheus/client_java/blob/f3f69858f84fb447732bbb33534e54be48f22afc/simpleclient_common/src/test/java/io/prometheus/client/exporter/common/TextFormatTest.java#L61

Official TextFormat has test for that. I'm not sure if we need to test that here.
I could create such test but it would require HTTP server mock which could slow down tests.

Reply:
This is the timestamp of the metrics sample not the timestamp of when the metrics was pushed to Puahgateway

Reply:
But in the output format you don't have 2 timestamps. Only one is supported. I checked that using following code:
```scala
import scala.collection.JavaConverters._

import java.io._
import io.prometheus.client.Collector
import java.util.Collections
import io.prometheus.client.exporter.common._
import com.banzaicloud.metrics.prometheus.client.exporter._

val mfs = Seq(
        new Collector.MetricFamilySamples(
                "Metric_Family_1",
                Collector.Type.GAUGE,
                "help1",
                Seq(
                        new Collector.MetricFamilySamples.Sample(
                                "Sample1",
                                Seq("label1", "label2").asJava,
                                Seq("1", "value1").asJava,
                                1.0,
                                3
                        ),
                        new Collector.MetricFamilySamples.Sample(
                                "Sample2",
                                Seq("label1", "label2").asJava,
                                Seq("2", "value2").asJava,
                                2.0,
                                3
                        )
                ).asJava
        )
).asJava


val writer1 = new StringWriter()
TextFormat.write004(writer, Collections.enumeration(mfs))
val writer2 = new StringWriter()
TextFormatWithTimestamp.write004(writer, Collections.enumeration(mfs), "3")

writer1.toString == writer2.toString
res16: Boolean = true
```

I use ConstantTimestampDecorator to override metric sample timestamp with provided push timestamp (if set)

Reply:
There are two types of timestamps:

1. The timestamp of a metric sample.
1. The timestamp when a collection of metrics are pushed to Pushgateway. This is different from the timestamp of metrics itself as it depends on the periodicity of the pushes (eg every 20 secs). The timestamp of when metrics are pushed to Pushgateway is written into the payload here: https://github.com/banzaicloud/spark-metrics/blob/master/src/main/java/com/banzaicloud/metrics/prometheus/client/exporter/TextFormatWithTimestamp.java#L106.

You say that with `ConstantTimestampDecorator` we push the timestamp of when the metric is passed to Pushgateway instead of the timestamp of the metric sample?

Reply:
@stoader I misinterpret the current code. I though that this second time of timestamp is static (instance creation time). I changed ConstantTimestampDecorator in the latest commit to use current time. So each time the `collect()` method is called it will override metrics sample timestamp.
As I understand 004 format in the request payload we always have only one timestamp. Only the interpretation of it may vary.

Reply:
Instead current time we should be able to pass in the push time such as to be the same across all metrics being pushed at the same time.